modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
  * limitations under the License.
  */
 
-// Chapel Strings
-module String {
+/*
+ * NOTES:
+ *
+ * - An empty string is represted by len == 0 and buff == nil.
+ *
+ * - It is assumed the bufferType is a local-only type, so we never
+ *   make a remote copy of one passed in by the user, though remote
+ *   copies are made of internal bufferType variables.
+ *
+ */
+
+//TODO: I dont think this module is overly useful anymore
+module BaseStringType {
   use CString;
+  use SysCTypes;
+
+  type bufferType = c_ptr(uint(8));
+  //param bufferTypeString: c_string = "c_ptr(uint(8))";
+  param chpl_string_min_alloc_size: int = 16;
+  // Growth factor to use when extending the buffer for appends
+  config param chpl_stringGrowthFactor = 1.5;
+  extern proc chpl_mem_goodAllocSize(minSize: size_t) : size_t;
+
+  extern type chpl_mem_descInt_t;
+
+  // We use this as a shortcut to get at here.id without actually constructing
+  // a locale object. Used when determining if we should make a remote transfer.
+  // This extern is also defined in ChapelLocale, but if it isn't defined here
+  // as well we will get a duplicate symbol error. I believe this is due to
+  // module load ordering issues.
+  // TODO: make a proc in ChapelLocale that returns this rather than using the extern
```

Comment:
I agree with this TODO. Also, I think we could have a primitive to check "is it local", which might simplify some of these cases. But you really want "is it within the same memory domain" rather than "is it on the same locale" (if there is NUMA locale model e.g.)